### ISSUE TITLE
FederatedDeployment with Volumes that include ConfigMap Items fail Validation

### ISSUE THREAD

**Issue by danehans** [kubernetes-sigs/kubefed#469]
When a federated deployment references a configmap that includes [item mapping](https://kubernetes.io/docs/concepts/storage/volumes/#configmap), it fails validation. For example:
```
volumes:
  - name: inject-config
    configMap:
      name: istio-sidecar-injector
      items:
        - key: config
          path: config
```

This worked recently, but now fails on master. Here is the example Deployment (succeeds):
```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: istio-sidecar-injector
  namespace: istio-system
  labels:
    app: sidecarInjectorWebhook
    chart: sidecarInjectorWebhook-1.0.3
    release: istio
    heritage: Tiller
    istio: sidecar-injector
spec:
  replicas: 1
  template:
    metadata:
      labels:
        istio: sidecar-injector
      annotations:
        sidecar.istio.io/inject: "false"
        scheduler.alpha.kubernetes.io/critical-pod: ""
    spec:
      serviceAccountName: istio-sidecar-injector-service-account
      containers:
        - name: sidecar-injector-webhook
          image: "docker.io/istio/sidecar_injector:1.0.3"
          imagePullPolicy: IfNotPresent
          args:
            - --caCertFile=/etc/istio/certs/root-cert.pem
            - --tlsCertFile=/etc/istio/certs/cert-chain.pem
            - --tlsKeyFile=/etc/istio/certs/key.pem
            - --injectConfig=/etc/istio/inject/config
            - --meshConfig=/etc/istio/config/mesh
            - --healthCheckInterval=2s
            - --healthCheckFile=/health
          volumeMounts:
          - name: config-volume
            mountPath: /etc/istio/config
            readOnly: true
          - name: certs
            mountPath: /etc/istio/certs
            readOnly: true
          - name: inject-config
            mountPath: /etc/istio/inject
            readOnly: true
          livenessProbe:
            exec:
              command:
                - /usr/local/bin/sidecar-injector
                - probe
                - --probe-path=/health
                - --interval=4s
            initialDelaySeconds: 4
            periodSeconds: 4
          readinessProbe:
            exec:
              command:
                - /usr/local/bin/sidecar-injector
                - probe
                - --probe-path=/health
                - --interval=4s
            initialDelaySeconds: 4
            periodSeconds: 4
          resources:
            requests:
              cpu: 10m
            
      volumes:
      - name: config-volume
        configMap:
          name: istio
      - name: certs
        secret:
          secretName: istio.istio-sidecar-injector-service-account
      - name: inject-config
        configMap:
          name: istio-sidecar-injector
          items:
          - key: config
            path: config
      affinity:      
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: beta.kubernetes.io/arch
                operator: In
                values:
                - amd64
                - ppc64le
                - s390x
          preferredDuringSchedulingIgnoredDuringExecution:
          - weight: 2
            preference:
              matchExpressions:
              - key: beta.kubernetes.io/arch
                operator: In
                values:
                - amd64
          - weight: 2
            preference:
              matchExpressions:
              - key: beta.kubernetes.io/arch
                operator: In
                values:
                - ppc64le
          - weight: 2
            preference:
              matchExpressions:
              - key: beta.kubernetes.io/arch
                operator: In
                values:
                - s390x
```

and the federated equivalent (fails):
```
apiVersion: primitives.federation.k8s.io/v1alpha1
kind: FederatedDeployment
metadata:
  name: istio-sidecar-injector
  namespace: istio-system
  labels:
    app: sidecarInjectorWebhook
    chart: sidecarInjectorWebhook-1.0.3
    release: istio
    heritage: Tiller
    istio: sidecar-injector
spec:
  template:
    spec:
      replicas: 1
      selector:
        matchLabels:
          istio: sidecar-injector
      template:
        metadata:
          labels:
            istio: sidecar-injector
          annotations:
            sidecar.istio.io/inject: "false"
            scheduler.alpha.kubernetes.io/critical-pod: ""
        spec:
          serviceAccountName: istio-sidecar-injector-service-account
          containers:
            - name: sidecar-injector-webhook
              image: "docker.io/istio/sidecar_injector:1.0.3"
              imagePullPolicy: IfNotPresent
              args:
                - --caCertFile=/etc/istio/certs/root-cert.pem
                - --tlsCertFile=/etc/istio/certs/cert-chain.pem
                - --tlsKeyFile=/etc/istio/certs/key.pem
                - --injectConfig=/etc/istio/inject/config
                - --meshConfig=/etc/istio/config/mesh
                - --healthCheckInterval=2s
                - --healthCheckFile=/health
              volumeMounts:
                - name: config-volume
                  mountPath: /etc/istio/config
                  readOnly: true
                - name: certs
                  mountPath: /etc/istio/certs
                  readOnly: true
                - name: inject-config
                  mountPath: /etc/istio/inject
                  readOnly: true
              livenessProbe:
                exec:
                  command:
                    - /usr/local/bin/sidecar-injector
                    - probe
                    - --probe-path=/health
                    - --interval=4s
                initialDelaySeconds: 4
                periodSeconds: 4
              readinessProbe:
                exec:
                  command:
                    - /usr/local/bin/sidecar-injector
                    - probe
                    - --probe-path=/health
                    - --interval=4s
                initialDelaySeconds: 4
                periodSeconds: 4
              resources:
                requests:
                  cpu: 10m
          volumes:
          - name: config-volume
            configMap:
              name: istio
          - name: certs
            secret:
              secretName: istio.istio-sidecar-injector-service-account
          - name: inject-config
            configMap:
              name: istio-sidecar-injector
              items:
                - key: config
                  path: config
          affinity:
            nodeAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
                nodeSelectorTerms:
                  - matchExpressions:
                      - key: beta.kubernetes.io/arch
                        operator: In
                        values:
                          - amd64
                          - ppc64le
                          - s390x
              preferredDuringSchedulingIgnoredDuringExecution:
                - weight: 2
                  preference:
                    matchExpressions:
                      - key: beta.kubernetes.io/arch
                        operator: In
                        values:
                          - amd64
                - weight: 2
                  preference:
                    matchExpressions:
                      - key: beta.kubernetes.io/arch
                        operator: In
                        values:
                          - ppc64le
                - weight: 2
                  preference:
                    matchExpressions:
                      - key: beta.kubernetes.io/arch
                        operator: In
                        values:
                          - s390x
---
apiVersion: primitives.federation.k8s.io/v1alpha1
kind: FederatedDeploymentPlacement
metadata:
  name: istio-sidecar-injector
  namespace: istio-system
  labels:
    app: sidecarInjectorWebhook
    chart: sidecarInjectorWebhook-1.0.3
    release: istio
    heritage: Tiller
    istio: sidecar-injector
spec:
  clusterNames:
  - cluster2
  - cluster1
```

Full error when creating the above example federated deployment:
```
The FederatedDeployment "istio-sidecar-injector" is invalid: []: Invalid value: map[string]interface {}{"spec":map[string]interface {}{"template":map[string]interface {}{"spec":map[string]interface {}{"replicas":1, "selector":map[string]interface {}{"matchLabels":map[string]interface {}{"istio":"sidecar-injector"}}, "template":map[string]interface {}{"metadata":map[string]interface {}{"annotations":map[string]interface {}{"scheduler.alpha.kubernetes.io/critical-pod":"", "sidecar.istio.io/inject":"false"}, "labels":map[string]interface {}{"istio":"sidecar-injector"}}, "spec":map[string]interface {}{"affinity":map[string]interface {}{"nodeAffinity":map[string]interface {}{"preferredDuringSchedulingIgnoredDuringExecution":[]interface {}{map[string]interface {}{"weight":2, "preference":map[string]interface {}{"matchExpressions":[]interface {}{map[string]interface {}{"key":"beta.kubernetes.io/arch", "operator":"In", "values":[]interface {}{"amd64"}}}}}, map[string]interface {}{"preference":map[string]interface {}{"matchExpressions":[]interface {}{map[string]interface {}{"key":"beta.kubernetes.io/arch", "operator":"In", "values":[]interface {}{"ppc64le"}}}}, "weight":2}, map[string]interface {}{"preference":map[string]interface {}{"matchExpressions":[]interface {}{map[string]interface {}{"key":"beta.kubernetes.io/arch", "operator":"In", "values":[]interface {}{"s390x"}}}}, "weight":2}}, "requiredDuringSchedulingIgnoredDuringExecution":map[string]interface {}{"nodeSelectorTerms":[]interface {}{map[string]interface {}{"matchExpressions":[]interface {}{map[string]interface {}{"key":"beta.kubernetes.io/arch", "operator":"In", "values":[]interface {}{"amd64", "ppc64le", "s390x"}}}}}}}}, "containers":[]interface {}{map[string]interface {}{"resources":map[string]interface {}{"requests":map[string]interface {}{"cpu":"10m"}}, "volumeMounts":[]interface {}{map[string]interface {}{"mountPath":"/etc/istio/config", "name":"config-volume", "readOnly":true}, map[string]interface {}{"mountPath":"/etc/istio/certs", "name":"certs", "readOnly":true}, map[string]interface {}{"mountPath":"/etc/istio/inject", "name":"inject-config", "readOnly":true}}, "args":[]interface {}{"--caCertFile=/etc/istio/certs/root-cert.pem", "--tlsCertFile=/etc/istio/certs/cert-chain.pem", "--tlsKeyFile=/etc/istio/certs/key.pem", "--injectConfig=/etc/istio/inject/config", "--meshConfig=/etc/istio/config/mesh", "--healthCheckInterval=2s", "--healthCheckFile=/health"}, "image":"docker.io/istio/sidecar_injector:1.0.3", "imagePullPolicy":"IfNotPresent", "livenessProbe":map[string]interface {}{"periodSeconds":4, "exec":map[string]interface {}{"command":[]interface {}{"/usr/local/bin/sidecar-injector", "probe", "--probe-path=/health", "--interval=4s"}}, "initialDelaySeconds":4}, "name":"sidecar-injector-webhook", "readinessProbe":map[string]interface {}{"exec":map[string]interface {}{"command":[]interface {}{"/usr/local/bin/sidecar-injector", "probe", "--probe-path=/health", "--interval=4s"}}, "initialDelaySeconds":4, "periodSeconds":4}}}, "serviceAccountName":"istio-sidecar-injector-service-account", "volumes":[]interface {}{map[string]interface {}{"configMap":map[string]interface {}{"name":"istio"}, "name":"config-volume"}, map[string]interface {}{"name":"certs", "secret":map[string]interface {}{"secretName":"istio.istio-sidecar-injector-service-account"}}, map[string]interface {}{"configMap":map[string]interface {}{"items":[]interface {}{map[string]interface {}{"key":"config", "path":"config"}}, "name":"istio-sidecar-injector"}, "name":"inject-config"}}}}}}}, "apiVersion":"primitives.federation.k8s.io/v1alpha1", "kind":"FederatedDeployment", "metadata":map[string]interface {}{"name":"istio-sidecar-injector", "namespace":"istio-system", "creationTimestamp":"2018-11-30T20:36:13Z", "labels":map[string]interface {}{"release":"istio", "app":"sidecarInjectorWebhook", "chart":"sidecarInjectorWebhook-1.0.3", "heritage":"Tiller", "istio":"sidecar-injector"}, "generation":1, "uid":"94360aab-f4df-11e8-85be-42010a8a0002", "selfLink":"", "clusterName":""}}: validation failure list:
type in spec.template.spec.template.spec.volumes.configMap is required
```

**Comment by danehans** [comment]
The workaround in the above k8s issue worked for me. Thanks @marun for discovering the issue.

**Comment by danehans** [comment]
Federating mutatingwebhookconfiguration is also effected by this issue:
```
apiVersion: primitives.federation.k8s.io/v1alpha1
kind: FederatedMutatingWebhookConfiguration
metadata:
  name: istio-sidecar-injector
  namespace: istio-system
  labels:
    app: istio-sidecar-injector
    chart: sidecarInjectorWebhook-0.8.0
    release: istio
    heritage: Tiller
spec:
  template:
    webhooks:
      - name: sidecar-injector.istio.io
        clientConfig:
          service:
            name: istio-sidecar-injector
            namespace: istio-system
            path: "/inject"
          caBundle: ""
        rules:
          - operations: [ "CREATE" ]
            apiGroups: [""]
            apiVersions: ["v1"]
            resources: ["pods"]
        failurePolicy: Fail
        namespaceSelector:
          matchLabels:
            istio-injection: enabled
```

Error:
```
The FederatedMutatingWebhookConfiguration "istio-sidecar-injector" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"primitives.federation.k8s.io/v1alpha1", "kind":"FederatedMutatingWebhookConfiguration", "metadata":map[string]interface {}{"namespace":"", "generation":1, "uid":"d9e0735e-fd01-11e8-8928-42010a8a006f", "selfLink":"", "clusterName":"", "name":"istio-sidecar-injector", "creationTimestamp":"2018-12-11T05:01:42Z", "labels":map[string]interface {}{"chart":"sidecarInjectorWebhook-0.8.0", "heritage":"Tiller", "release":"istio", "app":"istio-sidecar-injector"}}, "spec":map[string]interface {}{"template":map[string]interface {}{"webhooks":[]interface {}{map[string]interface {}{"clientConfig":map[string]interface {}{"caBundle":"", "service":map[string]interface {}{"name":"istio-sidecar-injector", "namespace":"istio-system", "path":"/inject"}}, "failurePolicy":"Fail", "name":"sidecar-injector.istio.io", "namespaceSelector":map[string]interface {}{"matchLabels":map[string]interface {}{"istio-injection":"enabled"}}, "rules":[]interface {}{map[string]interface {}{"apiVersions":[]interface {}{"v1"}, "operations":[]interface {}{"CREATE"}, "resources":[]interface {}{"pods"}, "apiGroups":[]interface {}{""}}}}}}}}: validation failure list:
spec.template.webhooks.clientConfig.caBundle in body must be of type byte: ""
```

**Comment by marun** [comment]
@danehans That appears to be a different problem - it doesn't think the value of `caBundle` is of type byte.  Maybe try skipping the quotes?

**Comment by marun** [comment]
It appears that this will be fixed by Kube 1.16,.

**Comment by lilic** [comment]
> It appears that this will be fixed by Kube 1.16,.

@marun any chance of this being backported to 1.15?
